File: Frontend/apps/astrogame_app/lib/models/researches/research.dart
import 'package:astrogame_app/communications/converters/guid_converter.dart';
import 'package:astrogame_app/communications/converters/research_converter.dart';
import 'package:astrogame_app/models/enums/research_type.dart';
import 'package:astrogame_app/models/technologies/technology.dart';
import 'package:json_annotation/json_annotation.dart';

@GuidConverter()
abstract class Research extends Technology {
  @JsonKey()
  late ResearchType researchType;

  @JsonKey()
  late double buildingTimeMultiplier;

  @JsonKey()
  late double buildingCostMultiplier;

  @JsonKey()
  late double buildingProductionMultiplier;

  @JsonKey()
  late double buildingConsumptionMultiplier;

  @JsonKey()
  late double structuralIntegrityMultiplier;

  @JsonKey()
  late double shieldPowerMultiplier;

  @JsonKey()
  late double weaponPowerMultiplier;

  @JsonKey()
  late double cargoCapacityMultiplier;

  @JsonKey()
  late double stellarSpeedMultiplier;

  @JsonKey()
  late double interstellarSpeedMultiplier;

  @JsonKey()
  late double fuelConsumptionMultiplier;

  Research();

  factory Research.fromJson(Map<String, dynamic> json) =>
      ResearchConverter().fromJson(json);
}

File: Backend/AstroGame.Shared/Models/Researches/Research.cs
using System.Collections.Generic;
using System.ComponentModel;

namespace AstroGame.Shared.Models.Researches
{
    public abstract class Research : Technology
    {
        public string Name { get; set; }
        public string Description { get; set; }
        public int Order { get; set; }
        public string AssetName { get; set; }
        public ResearchType ResearchType { get; set; }

        [DefaultValue(1)] public double BuildingTimeMultiplier { get; set; }
        [DefaultValue(1)] public double BuildingCostMultiplier { get; set; }
        [DefaultValue(1)] public double BuildingProductionMultiplier { get; set; }
        [DefaultValue(1)] public double BuildingConsumptionMultiplier { get; set; }

        [DefaultValue(1)] public double StructuralIntegrityMultiplier { get; set; }
        [DefaultValue(1)] public double ShieldPowerMultiplier { get; set; }
        [DefaultValue(1)] public double WeaponPowerMultiplier { get; set; }
        [DefaultValue(1)] public double CargoCapacityMultiplier { get; set; }
        [DefaultValue(1)] public double StellarSpeedMultiplier { get; set; }
        [DefaultValue(1)] public double InterstellarSpeedMultiplier { get; set; }
        [DefaultValue(1)] public double FuelConsumptionMultiplier { get; set; }

        public virtual List<ResearchCost> ResearchCosts { get; set; }
    }
}